File: DummyClient/Lock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

public class Lock
{
    const long ACQUIRE_TIMEOUT_TICK         = 10000;
    const uint MAX_SPIN_COUNT               = 5000;
    const uint WRITE_THREAD_MASK            = 0xFFFF0000;
    const uint READ_COUNT_MASK              = 0x0000FFFF;
    const uint EMPTY_FLAG                   = 0x0000000;

    uint _lockFlag                          = EMPTY_FLAG;
    ushort _writeCount                      = 0;

    public void WriteLock()
    {
        uint lockThreadId = _lockFlag >> 16;
        if(Thread.CurrentThread.ManagedThreadId == lockThreadId)
        {
            _writeCount++;
            return;
        }

        long beginTick = Environment.TickCount64;
        uint desired = ((uint)Thread.CurrentThread.ManagedThreadId << 16) & WRITE_THREAD_MASK;
        while (true)
        {
            for (uint spinCount = 0; spinCount < MAX_SPIN_COUNT; spinCount++)
            {
                if (Interlocked.CompareExchange(ref _lockFlag, desired, EMPTY_FLAG) == EMPTY_FLAG)
                {
                    _writeCount++;
                    return;
                }
            }

            if (Environment.TickCount64 - beginTick > ACQUIRE_TIMEOUT_TICK)
                Environment.Exit(0);

            Thread.Yield();
        }
    }
    public void WriteUnLock()
    {
        if ((_lockFlag & READ_COUNT_MASK) != 0)
            Environment.Exit(0);

        ushort lockCount = --_writeCount;
        if (lockCount == 0)
            Interlocked.Exchange(ref _lockFlag, EMPTY_FLAG);
    }
    public void ReadLock()
    {
        ulong lockThreadId = (_lockFlag & WRITE_THREAD_MASK) >> 16;
        if (lockThreadId == (ulong)Thread.CurrentThread.ManagedThreadId)
        {
            Interlocked.Increment(ref _lockFlag);
            return;
        }

        long beginTick = Environment.TickCount64;
        while (true)
        {
            for (uint spinCount = 0; spinCount < MAX_SPIN_COUNT; spinCount++)
            {
                uint expected = _lockFlag & READ_COUNT_MASK;

                if (Interlocked.CompareExchange(ref _lockFlag, expected + 1, expected) == expected)
                    return;
            }

            if (Environment.TickCount64 - beginTick > ACQUIRE_TIMEOUT_TICK)
                Environment.Exit(0);

            Thread.Yield();
        }
    }
    public void ReadUnLock()
    {
        if ((Interlocked.Decrement(ref _lockFlag) & READ_COUNT_MASK) == 0)
            Environment.Exit(0);
    }
}

public class ReadLock : IDisposable
{
    private Lock _lock;

    public ReadLock(Lock rwLock)
    {
        _lock = rwLock;
        _lock.ReadLock();
    }

    public void Dispose()
    {
        _lock.ReadUnLock();
    }
}

public class WriteLock : IDisposable
{
    private Lock _lock;

    public WriteLock(Lock rwLock)
    {
        _lock = rwLock;
        _lock.WriteLock();
    }

    public void Dispose()
    {
        _lock.WriteUnLock();
    }
}

File: ServerCore/Lock.cpp
#include "pch.h"
#include "Lock.h"
#include "DeadLockProfiler.h"

void Lock::WriteLock(const char* name)
{
#if DEV_TEST
	DeadLockProfiler::GetInstance().PushLock(name);
#endif

	const uint32 lockThreadId = (_lockFlag.load() & WRITE_THREAD_MASK) >> 16;
	if (LThreadId == lockThreadId)
	{
		_writeCount++;
		return;
	}

	const uint64 beginTick = ::GetTickCount64();
	const uint32 desired = (LThreadId << 16) & WRITE_THREAD_MASK;
	while (true)
	{
		for (uint32 spinCount = 0; spinCount < MAX_SPIN_COUNT; spinCount++)
		{
			uint32 expected = EMPTY_FLAG;
			if (_lockFlag.compare_exchange_strong(OUT expected, desired))
			{
				_writeCount++;
				return;
			}
		}

		if (::GetTickCount64() - beginTick > ACQUIRE_TIMEOUT_TICK)
			CRASH("LOCK_TIMEOUT");

		std::this_thread::yield();
	}
}

void Lock::WriteUnlock(const char* name)
{
#if DEV_TEST
	DeadLockProfiler::GetInstance().PopLock(name);
#endif

	if ((_lockFlag.load() & READ_COUNT_MASK) != 0)
		CRASH("INVALID_UNLOCK_ORDER");

	const int32 lockCount = --_writeCount;
	if (lockCount == 0)
		_lockFlag.store(EMPTY_FLAG);
}

void Lock::ReadLock(const char* name)
{
#if DEV_TEST
	DeadLockProfiler::GetInstance().PushLock(name);
#endif

	const uint32 lockThreadId = (_lockFlag.load() & WRITE_THREAD_MASK) >> 16;
	if (lockThreadId == LThreadId)
	{
		_lockFlag++;
		return;
	}

	const uint64 beginTick = ::GetTickCount64();
	while (true)
	{
		for (uint32 spinCount = 0; spinCount < MAX_SPIN_COUNT; spinCount++)
		{
			uint32 expected = (_lockFlag.load() & READ_COUNT_MASK);
			if (_lockFlag.compare_exchange_strong(OUT expected, expected + 1))
				return;
		}

		if (::GetTickCount64() - beginTick > ACQUIRE_TIMEOUT_TICK)
			CRASH("LOCK_TIMEOUT");

		std::this_thread::yield();
	}
}

void Lock::ReadUnlock(const char* name)
{
#if DEV_TEST
	DeadLockProfiler::GetInstance().PopLock(name);
#endif

	if ((_lockFlag.fetch_sub(1) & READ_COUNT_MASK) == 0)
		CRASH("MULTIPLE_UNLOCK");
}
